### PR TITLE
Add a model builder to ease test creation

### PR DESCRIPTION
**Pull request by antoine-de**
linked to #176 

first working draft to a `Model` builder to ease unit test creation.

It is an early draft, the implementation is lacking in several ways for the moment, I suggest a first review on the final API before reviewing the code (there are lot of stuff to discuss on it :roll_eyes: : ).

To see the api, the unittest is a good starting point.
it looks like:

```rust
let model = ModelBuilder::default()
    .vj("toto", |vj| {
        vj.st("A", "10:00:00", "10:01:00")
          .st("B", "11:00:00", "11:01:00");
    })
    .vj("tata", |vj| {
        vj.st("A", "10:00:00", "10:01:00")
          .st("D", "11:00:00", "11:01:00");
    })
    .build();
```

I'm not a big lambda fan, but it seems the only way to "nest" a builder pattern.

Note: the builder return itself by value which can lead to  some twerk if not chaining the method, but this way the minimal example is quite simple.

Note: to control the stop_point/stop_area creation we could also add a `.sa(...)`method to the `ModelBuilder`

What do you think about this ?

late Node: I'm in holidays tonight, you have some time to think about this :stuck_out_tongue_winking_eye:

### CODE REVIEW

**Comment by antoine-de** [comment]
pouet, any thought on this ?

I'm not convinced at all with the way to initialize the route (and even more for the line, network, ....) of the VJ, but honestly since I don't need it for my use case, I think I'll keep it simple for the moment

**Comment by TeXitoi** [comment]
My point would be "go for it, we can refactor later if needed."

@datanel @nberard what do you think?

**Comment by datanel** [comment]
yes, if it is just for test (currently) it is not so critical so have fun and we refactor later.

**Comment by antoine-de** [comment]
ok, nice, I'll remove the ` #![allow(missing_docs)]` and some minor cleanup (I think I'll put the `WithId` behind a macro), but I think you can start reviewing the code ?

**Comment by TeXitoi** [comment]
If we want to publish on crates.io, maybe better to do a dedicated sub crate to have a different semver on that.

**Comment by antoine-de** [comment]
since this subcrate would depend on navitia_model how do you propose to make this link ?

**Comment by antoine-de** [comment]
I made all the modification I planned to do, so I think you can review it

File: src/objects.rs [hunk 1]
```diff
@@ -740,6 +784,16 @@ impl FromStr for Time {
...
 }
+
+// non failing conversion from String to Time
+// if the result has to be checked, it's better to use `parse()`
+// mainly used for the ModelBuilder
+impl From<&str> for Time {
+    fn from(s: &str) -> Time {
+        s.parse().unwrap()
+    }
+}
```
[the remark this answers is not in the document]
Comment (TeXitoi) [reply src/objects.rs]: What I don't like is that this impl is public for any duration, not only the builder need.

You can define a IntoDuration trait and implement it for &str, &Duration and Duration as a workaround, specifying that IntoDuration might fail.

File: src/objects.rs [hunk 2]
```diff
@@ -122,6 +122,22 @@ macro_rules! impl_comment_links {
...
+pub trait WithId {
+    fn with_id(id: &str) -> Self;
+}
+
+macro_rules! impl_with_id {
+    ($ty:ty) => {
+        impl WithId for $ty {
+            fn with_id(id: &str) -> Self {
+                let mut r = Self::default();
+                r.id = id.to_owned();
```
[the remark this answers is not in the document]
Comment (antoine-de) [reply src/objects.rs]: hum for this, how do you think it should change the builder ?

should the builder always ask for a name before creating an object ? or take an Option ?

`builder.route("1", "route 1")`  (I'm not a fan, it makes the simple example less simple)
or `builder.route("1", Some("route 1"))` (not a fan too, too complex)

or just set the name as the id by default (but then I don't really see the point of changing this Trait.

[the remark this answers is not in the document]
Comment (TeXitoi) [reply src/objects.rs]: The point was to do the macro like that:
```rust
        impl WithId for $ty {
            fn with_id(id: &str) -> Self {
                let mut r = Self::default();
                r.id = id.to_owned();
                r.name = id.to_owned();
                r
            }
        }
```

like that, if you ask for
```rust
model = builder.vj("vj1", |vj| vj.route("r1")).build();
```
the name of the created route will be useful.

**Comment by antoine-de** [comment]
thanks for the review (and for the help!), I think I changed everything

File: model-builder/src/builder.rs [hunk 3]
```diff
@@ -115,13 +108,36 @@ impl<'a> ModelBuilder {
...
             model: &mut self,
             vj_idx,
-            route_modifier: None,
         };
 
         vj_initer(vj_builder);
         self
     }
 
+    /// Add a new VehicleJourney to the model
```
Comment (TeXitoi) [inline model-builder/src/builder.rs]: bad comment

File: model-builder/src/builder.rs [hunk 4]
```diff
@@ -115,13 +108,36 @@ impl<'a> ModelBuilder {
...
+    /// let model = model_builder::ModelBuilder::default()
+    ///        .route("l1", |r| {
+    ///             r.name = "ligne 1".to_owned();
+    ///         })
+    ///      .vj("toto", |vj| {
+    ///          vj.route("l1")
+    ///            .st("A", "10:00:00", "10:01:00")
+    ///            .st("B", "11:00:00", "11:01:00");
+    ///      })
+    ///        .build();
```
Comment (TeXitoi) [inline model-builder/src/builder.rs]: bad indentation

File: model-builder/src/builder.rs [hunk 5]
```diff
@@ -169,7 +185,7 @@ impl<'a> VehicleJourneyBuilder<'a> {
                 get_or_create_with(
                     &mut self.model.collections.stop_areas,
                     &sa_id,
-                    Some(|mut sa: &mut crate::objects::StopArea| sa.name = format!("sa {}", sp)),
+                    |mut sa: &mut crate::objects::StopArea| sa.name = format!("sa {}", sp),
```
Comment (TeXitoi) [inline model-builder/src/builder.rs]: That's now useless, no?

File: model-builder/src/builder.rs [hunk 6]
```diff
@@ -226,23 +242,24 @@ impl<'a> VehicleJourneyBuilder<'a> {
...
-            id: name.to_owned(),
-            modifier: Box::new(f),
-        });
+    pub fn route(self, id: &str) -> Self {
+        {
+            let vj = &mut self
+                .model
+                .collections
+                .vehicle_journeys
+                .index_mut(self.vj_idx);
```
Comment (TeXitoi) [inline model-builder/src/builder.rs]: you can use `.vehicle_journeys[self.vj_idx];`

File: model-builder/src/builder.rs [hunk 7]
```diff
@@ -252,21 +269,14 @@ impl<'a> Drop for VehicleJourneyBuilder<'a> {
     fn drop(&mut self) {
         let collections = &mut self.model.collections;
         // add the missing objects to the model (routes, lines, ...)
-        let mut new_vj = collections.vehicle_journeys.index_mut(self.vj_idx);
+        let new_vj = collections.vehicle_journeys.index_mut(self.vj_idx);
```
Comment (TeXitoi) [inline model-builder/src/builder.rs]: you can use .vehicle_journeys[self.vj_idx];

File: model-builder/src/builder.rs (hunk 6 above)
Comment (antoine-de) [reply model-builder/src/builder.rs]: I don't think so, I need a mutable vj here.

Comment (TeXitoi) [reply model-builder/src/builder.rs]: Yeah, my bad, `index_mut` on `CollectionWithId` is custom.

**Pull request merged**